Model/Api/OAuthRedirect.php

```diff
@@ -35,37 +46,209 @@ class OAuthRedirect implements OAuthRedirectInterface
+     * @param CustomerInterfaceFactory         $customerInterfaceFactory
+     * @param CustomerFactory                  $customerFactory
+     * @param Url                              $url
+     * @param Bugsnag                          $bugsnag
      */
     public function __construct(
         Response $response,
-        DeciderHelper $deciderHelper
+        DeciderHelper $deciderHelper,
+        SSOHelper $ssoHelper,
+        ExternalCustomerEntityRepository $externalCustomerEntityRepository,
+        CustomerRepository $customerRepository,
+        StoreManagerInterface $storeManager,
+        CustomerSession $customerSession,
+        CustomerInterfaceFactory $customerInterfaceFactory,
+        CustomerFactory $customerFactory,
+        Url $url,
+        Bugsnag $bugsnag
     ) {
         $this->response = $response;
         $this->deciderHelper = $deciderHelper;
+        $this->ssoHelper = $ssoHelper;
+        $this->externalCustomerEntityRepository = $externalCustomerEntityRepository;
+        $this->customerRepository = $customerRepository;
+        $this->storeManager = $storeManager;
+        $this->customerSession = $customerSession;
+        $this->customerInterfaceFactory = $customerInterfaceFactory;
+        $this->customerFactory = $customerFactory;
+        $this->url = $url;
+        $this->bugsnag = $bugsnag;
     }
 
     /**
      * Login with Bolt SSO and redirect
      *
      * @api
      *
-     * @param string $authorization_code
+     * @param string $code
+     * @param string $scope
+     * @param string $state
      *
      * @return void
      *
      * @throws NoSuchEntityException
+     * @throws WebapiException
      */
-    public function login($authorization_code = '')
+    public function login($code = '', $scope = '', $state = '')
     {
         if (!$this->deciderHelper->isBoltSSOEnabled()) {
             throw new NoSuchEntityException(__('Request does not match any route.'));
         }
 
-        $this->response->setHeader('Content-Type', 'application/json');
-        $this->response->setHttpResponseCode(500);
-        $this->response->setBody(json_encode(['message' => 'Not implemented']));
-        $this->response->sendResponse();
+        if ($code === '' || $scope === '' || $state === '') {
+            throw new WebapiException(__('Bad Request'), 0, WebapiException::HTTP_BAD_REQUEST);
+        }
+
+        list($clientID, $clientSecret, $boltPublicKey) = $this->ssoHelper->getOAuthConfiguration();
+
+        $token = $this->ssoHelper->exchangeToken($code, $scope, $clientID, $clientSecret);
+        if (is_string($token)) {
+            $this->bugsnag->notifyError('OAuthRedirect', $token);
+            throw new WebapiException(__('Internal Server Error'), 0, WebapiException::HTTP_INTERNAL_ERROR);
+        }
+
+        $payload = $this->ssoHelper->parseAndValidateJWT($token, $clientID, $boltPublicKey);
+        if (is_string($payload)) {
```

Comment:
why is the check `is_string`?

Reply:
because i'm returning an error message if there's an error, otherwise an object